tutorials/graph-theory/prims-algorithm.md

```diff
@@ -1,12 +1,378 @@
+
+### Approach
+First, we convert the given 'connections' array, into an adjacency list for ease of traversing.
+<Tabs>
+
+<TabItem value="py" label="Python">
+<SolutionAuthor name="@Infonioknight"/>
+  
+```py
+connections = [[1,2,5],[1,3,6],[2,3,1]] # Example
+adjacency_list = {}
+
+for connection in connections:
+    if connection[0] not in adjacency_list:
+        adjacency_list[connection[0]] = []
+    if connection[1] not in adjacency_list:
+        adjacency_list[connection[1]] = []
+    adjacency_list[connection[0]].append((connection[2], connection[1]))
+    adjacency_list[connection[1]].append((connection[2], connection[0]))
+```
+</TabItem>
+</Tabs>
+
+The adjacency list will look like:
+```
+1: [(5, 2), (6, 3)]
+2: [(5, 1), (1, 3)]
+3: [(6, 1), (1, 2)]
+```
+Where each tuple consists of '(cost, destination)'.
+
+Now, to find the minimum cost, we use the below approach.
+### Explanation:
+
+<Tabs>
+<TabItem value="py" label="Python">
+<SolutionAuthor name="@Infonioknight"/>
+
+```py
+for connection in adjacency_list[start]:
+    heapq.heappush(connection_queue, connection)
+```
+</TabItem>
+</Tabs>
+
+First, we push all the connections of the starting node into the priority queue. On doing so, the nodes with edges of MINIMUM COST come in front.
+
+<Tabs>
+<TabItem value="py" label="Python">
+<SolutionAuthor name="@Infonioknight"/>
+  
+```py
+while len(seen) < n and connection_queue:
+    cost, current = heapq.heappop(connection_queue)
+    if current in seen:
+        continue
+
+    res += cost
+    seen.add(current)
+```
+</TabItem>
+</Tabs>
+
+- Then, we take the first element out of the queue (this is the node we ideally will add to our MST as it will have the minimum cost). We also ensure that the node selected hasn't already been visited.
+- If the node hasn't been visited, we then add the cost to our solution and mark the node as visited.
+
+<Tabs>
+<TabItem value="py" label="Python">
+<SolutionAuthor name="@Infonioknight"/>
+  
+```py
+    for connection in adjacency_list[current]:
+        if connection[1] not in seen:
+            heapq.heappush(connection_queue, connection)
+
+if len(seen) < n: return -1
+else: return res
+```
+</TabItem>
+</Tabs>
+
+- After that, we check the adjacency list for all the nodes connected to the current node (the node visited most recently) and if the connections of this node haven't been visited, they are added to the priority queue.
+- This process goes on until one of two outcomes is achieved.
+    - **Not all** the nodes have been visited, in which case the number of nodes in 'seen' will be less than the total number of nodes. In this case, we do not have a successful output and return $-1$
+    - If all the nodes have been visited, we return the solution.
+
+#### Final Code:
+
+<Tabs>
+<TabItem value="py" label="Python">
+<SolutionAuthor name="@Infonioknight"/>
+  
+```py
+import random
+import heapq
+
+class Solution:
+  def connecting_cities_with_minimum_cost(self, adjacency_list):
```

Comment:
pls check the format in your localhost

<img width="887" alt="image" src="https://github.com/user-attachments/assets/984cda00-eb8b-443c-99b2-391e0f8f181e" />